css-ui-3/outline-004.html

```diff
@@ -0,0 +1,29 @@
+<!doctype html>
+<html lang=en>
+  <meta charset=utf-8>
+  <title>CSS-UI test: connected outline parts</title>
+  <link rel="author" href="http://florian.rivoal.net/">
+  <meta name=flags content="ahem">
+  <meta name=assert content="Each part of the outline should be fully connected rather than open on some sides">
+  <link rel="match" href="../reference/ref-filled-green-100px-square.xht">
+  <link rel=help href="https://drafts.csswg.org/css-ui-3/#outline-props">
+<style>
+span {
+  outline: solid green 10px;
+  color: green;
+}
+div {
+  box-sizing: border-box;
+  width: 100px;
+  height: 100px;
+  padding: 10px;
+  line-height: 40px;
+  font-size: 40px;
+  background: red;
+  font-family: ahem;
+}
+</style>
+
+  <p>Test passes if there is a filled green square and <strong>no red</strong>.</p>
+  <div><span>xx<br>xx</span></div>
```

Comment:
> Is this guaranteed to pass?

I think so.

> Where is it defined exactly where the outline is drawn? What if it's drawn 10px away from the border edge?

https://drafts.csswg.org/css-ui-3/#outline-offset
> By default, the outline is drawn starting just outside the border edge.

&

> If the computed value of outline-offset is anything other than 0, then the outline is outset from the border edge by that amount.

I agree that the spec is vague and gives a lot of liberties to implementers to do creative things to deal with all sorts of situations, but I don't think it is "anything goes" permissive, and the basic situations are expected to be simple and predictable.

My interpretation and working assumptions for many tests is basically:
1. if outline-style: auto, all bets are off, and the only thing you should be able to count on is that there is an outline.
2. if outline-style something other than auto (or none), the outline is predictable in simple cases, even though it is not in the general case: if there are not rounded corners, if there's no transform, if there's nothing overflowing from the box, if the box is not fragmented... then you can expect a simple line following the border edge.
3. If the box is fragmented, there may either be one outline per fragment (each following the 1. and 2.), or a outline single outline circumscribing all fragments. If the union of all fragments is a rectangle, then the single outline (if it is a single one) can be expected be a rectangle as well, but If the union of all fragments is a not rectangle, the shape of the outline is unpredictable, it may be the smallest bounding box, it may be an irregular shape fitting everything, who knows.
4. If outlines overlap with eachother or with various boxes, what what order they're painted is undefined, but that's not a license to change their shape, their color, their position, etc...

In the case of this particular test, it is designed to pass regardless of whether we end up with a separate outline for each line, or one that goes around both lines.

I may have made a few mistakes, but I've tried to be fairly careful to make sure the tests keep on passing when the implementations do take the liberties that the spec allows for. But if your interpretation of the spec is that implementations can do anything all the time, then sure, we cannot possibly write any test. But such a spec would not be terribly useful. As this matches implementations, I feel fairly comfortable doing so.